lib/collection/src/collection.rs

```diff
@@ -1173,7 +1162,52 @@ impl Collection {
                     }
                     drop(read_shard_holder);
 
-                    // 2 - mark peer as inactive
+                    // 2 - promote local or remote to replica set if needed
+                    if *peer_id == self.this_peer_id {
+                        let mut write_shard_holder = self.shards_holder.write().await;
+                        let shard_to_insert = match write_shard_holder.take_shard(*shard_id) {
+                            Some(Shard::Local(local_shard)) => {
+                                let new_replica = ReplicaSetShard::build_from_local(
+                                    *shard_id,
+                                    self.this_peer_id,
+                                    local_shard,
+                                    self.notify_peer_failure_cb.clone(),
+                                )
+                                .await?;
+                                Shard::ReplicaSet(new_replica)
+                            }
+                            Some(Shard::Remote(remote_shard)) => {
+                                // TODO is that the best way to retrieve the current path?
+                                let shard_path = latest_shard_paths(&self.path, *shard_id)
+                                    .await?
+                                    .first()
+                                    .unwrap() // TODO unwrap
+                                    .0
+                                    .clone();
+                                let new_replica = ReplicaSetShard::build_from_remote(
+                                    *shard_id,
+                                    &shard_path,
+                                    self.this_peer_id,
+                                    remote_shard,
+                                    self.notify_peer_failure_cb.clone(),
+                                )
+                                .await?;
+                                Shard::ReplicaSet(new_replica)
+                            }
+                            Some(other_shard) => other_shard, // shard needs to be reinserted again
```

Comment:
`shard needs to be reinserted again`
Why? I think we can just skip inserting it?

Return `None` here and add later:
```rust
if let Some(shard_to_insert) = shard_to_insert {
write_shard_holder.replace_shard(*shard_id, shard_to_insert);
}
```

Reply:
Because I am matching on 

```rust
match write_shard_holder.take_shard(*shard_id) {
...
```

Which removes the shard and returns ownership of. 

This is exactly what is needed for local and remote to build a new replica_set.

However, in other cases, the shard needs to be reinserted.

Do you see a better way here? I wanted to avoid doing first a read and then a write.